src/index.ts

```diff
@@ -95,129 +73,97 @@ declare module 'axios' {
     // Cannot determine if the request can be retried
     return false;
   }
 
   return isRetryableError(error) && SAFE_HTTP_METHODS.indexOf(error.config.method) !== -1;
 }
 
-/**
- * @param  {Error}  error
- * @return {boolean}
- */
-export function isIdempotentRequestError(error) {
-  if (!error.config) {
+export function isIdempotentRequestError(error: AxiosError) {
+  if (!error.config?.method) {
     // Cannot determine if the request can be retried
     return false;
   }
-
   return isRetryableError(error) && IDEMPOTENT_HTTP_METHODS.indexOf(error.config.method) !== -1;
 }
 
-/**
- * @param  {Error}  error
- * @return {boolean}
- */
-export function isNetworkOrIdempotentRequestError(error) {
+export function isNetworkOrIdempotentRequestError(error: AxiosError) {
   return isNetworkError(error) || isIdempotentRequestError(error);
 }
 
-/**
- * @return {number} - delay in milliseconds, always 0
- */
 function noDelay() {
   return 0;
 }
 
-/**
- * Set delayFactor 1000 for an exponential delay to occur on the order
- * of seconds
- * @param  {number} [retryNumber=0]
- * @param  {Error}  _error - unused; for existing API of retryDelay callback
- * @param  {number} [delayFactor=100] milliseconds
- * @return {number} - delay in milliseconds
- */
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-export function exponentialDelay(retryNumber = 0, _error = undefined, delayFactor = 100) {
+export function exponentialDelay(
+  retryNumber = 0,
+  _error: AxiosError | undefined = undefined,
+  delayFactor = 100
+) {
   const delay = 2 ** retryNumber * delayFactor;
   const randomSum = delay * 0.2 * Math.random(); // 0-20% of the delay
   return delay + randomSum;
 }
 
-/** @type {IAxiosRetryConfig} */
-export const DEFAULT_OPTIONS = {
+export const DEFAULT_OPTIONS: Required<IAxiosRetry.IAxiosRetryConfig> = {
   retries: 3,
   retryCondition: isNetworkOrIdempotentRequestError,
   retryDelay: noDelay,
   shouldResetTimeout: false,
   onRetry: () => {}
 };
 
-/**
- * Returns the axios-retry options for the current request
- * @param  {AxiosRequestConfig} config
- * @param  {IAxiosRetryConfig} defaultOptions
- * @return {IAxiosRetryConfigExtended}
- */
-function getRequestOptions(config, defaultOptions) {
+function getRequestOptions(
+  config: AxiosRequestConfig,
+  defaultOptions: IAxiosRetry.IAxiosRetryConfig
+): Required<IAxiosRetry.IAxiosRetryConfig> & IAxiosRetry.IAxiosRetryConfigExtended {
   return { ...DEFAULT_OPTIONS, ...defaultOptions, ...config[namespace] };
 }
 
-/**
- * Initializes and returns the retry state for the given request/config
- * @param  {AxiosRequestConfig} config
- * @param  {IAxiosRetryConfig} defaultOptions
- * @return {IAxiosRetryConfigExtended}
- */
-function getCurrentState(config, defaultOptions) {
-  const currentState = getRequestOptions(config, defaultOptions);
+function setCurrentState(
+  config: AxiosRequestConfig,
+  defaultOptions: IAxiosRetry.IAxiosRetryConfig | undefined
+) {
```

Comment:
Can you also add annotations for the return values? Those were present with JSDoc but are not anymore.

Not having those means that Typescript will infer it but that could easily then cause situation where the return value was supposed to be a boolean but some change will make it "boolean | undefined" and nothing will catch that. And that can be a breaking change.

Reply:
I agree on that, but only for the public types. I'll have a look. Thanks!